File: Session4/C++/LongestPalindromicSubstring.cpp
https://leetcode.com/problems/longest-palindromic-substring/

class Solution {  

    int maxLen = 1;
    int startIndex = 0;

    void longestPalindromicSubstring(int l, int r, string &s)
    {
        int n = s.size();
        while(l>=0 && r<n && s[l]==s[r]) {
            l--;
            r++;
        }
        int currLen = r-l-1;
        if(currLen > maxLen) {
            startIndex = l+1;
            maxLen = currLen;
        }
    }

public:
    string longestPalindrome(string s) {
        int n = s.size();
        
        for(int i=0; i<n; i++) {
            // odd length
            longestPalindromicSubstring(i, i, s);
            // even length
            longestPalindromicSubstring(i, i+1, s);
        }
        return s.substr(startIndex, maxLen);
    }
};

File: Session4/Java/LongestPalindromicSubstring.java
https://leetcode.com/problems/longest-palindromic-substring/

class Solution {

    private int maxLen = 1;
    private int startIndex = 0;

    private void longestPalindromicSubstring(int l, int r, String s) {
        int n = s.length();
        while (l >= 0 && r < n && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        int currLen = r - l - 1;
        if (currLen > maxLen) {
            maxLen = currLen;
            startIndex = l + 1;
        }
    }

    public String longestPalindrome(String s) {
        
        int n = s.length();
        
        for (int i = 0; i < n; i++) {
            // Odd length
            longestPalindromicSubstring(i, i, s);
            // Even length
            longestPalindromicSubstring(i, i + 1, s);
        }

        return s.substring(startIndex, startIndex + maxLen);
    }
}